core/drivers/crypto/hisilicon/sec_main.c

```diff
@@ -0,0 +1,265 @@
+{
+	struct hisi_qm *qm = &sec_dev->qm;
+	uint32_t val = 0;
+
+	if (qm->fun_type == HISI_QM_HW_VF)
+		return HISI_QM_DRVCRYPT_NO_ERR;
+
+	/* QM_HW_V2 version need to close clock gating */
+	io_clrbits32(qm->io_base + SEC_CONTROL_REG, SEC_CLK_GATE_ENABLE);
+
+	hisi_qm_dev_init(qm);
+
+	io_write32(qm->io_base + SEC_MEM_START_INIT, 0x1);
+
+	if (IO_READ32_POLL_TIMEOUT(qm->io_base + SEC_MEM_INIT_DONE, val,
+				   val & 0x1, POLL_PERIOD, POLL_TIMEOUT)) {
+		EMSG("Fail to init sec mem");
+		return HISI_QM_DRVCRYPT_ETMOUT;
+	}
+
+	io_setbits32(qm->io_base + SEC_CONTROL_REG, sec_dev->endian);
+
+	if (qm->version == HISI_QM_HW_V2) {
+		/* SMMU bypass */
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL0,
+			   SEC_USER0_CFG);
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL1,
+			   SEC_USER1_CFG);
+		io_write32(qm->io_base + AM_CFG_SINGLE_PORT_MAX_TRANS,
+			   SEC_SINGLE_PORT_MAX_TRANS);
+		io_write32(qm->io_base + SEC_SAA_EN, SEC_SAA_ENABLE);
+		/* HW V2 enable SM4 extra mode, as CTR/ECB */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN0, SEC_BD_ERR_CHK0);
+		/* Enable SM4 xts mode multiple iv */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN1, SEC_BD_ERR_CHK1);
+		/* disable PBKDF2 len check */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN2, SEC_BD_ERR_CHK2);
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN3, SEC_BD_ERR_CHK3);
+	} else {
+		/* cmd_type is controlled by HAC subctrl, default normal */
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL0_V3,
+			   SEC_USER0_CFG_V3);
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL1_V3,
+			   SEC_USER1_CFG_V3);
+		io_write32(qm->io_base + SEC_SAA_EN, SEC_SAA_ENABLE_V3);
+		/* disable PBKDF2 salt len check */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN2_V3,
+			   SEC_BD_ERR_CHK2);
+	}
+	io_write32(qm->io_base + SEC_RAS_CE_ENABLE, SEC_RAS_CE_ENB_MASK);
+	io_write32(qm->io_base + SEC_RAS_FE_ENABLE, SEC_RAS_FE_ENB);
+	io_write32(qm->io_base + SEC_RAS_NFE_ENABLE, SEC_RAS_NFE_ENB_MASK);
+	io_write32(qm->io_base + SEC_CORE_INT_MASK, SEC_ABNML_INT_DISABLE);
+
+	sec_enable_clock_gate(qm);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status sec_dev_status_check(struct hisi_qm *qm)
+{
+	uint32_t val = 0;
+
+	val = io_read32(qm->io_base + SEC_CORE_INT_SOURCE);
+	if (val & SEC_RAS_NFE_ENB_MASK) {
+		EMSG("SEC NFE RAS happened, need to reset");
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	val = io_read32(qm->io_base + HISI_QM_ABNML_INT_SRC);
+	if (val) {
+		if (val & HISI_QM_SEC_NFE_INT_MASK)
+			EMSG("QM NFE RAS happened, need to reset");
+
+		if (val & HISI_QM_INVALID_DB) {
+			EMSG("QM invalid db happened, please check");
+			io_write32(qm->io_base + HISI_QM_ABNML_INT_SRC,
+				   HISI_QM_INVALID_DB);
+		}
+
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status sec_qm_init(struct acc_device *sec_dev)
+{
+	struct hisi_qm *qm = &sec_dev->qm;
+
+	if (cpu_mmu_enabled()) {
```

Comment:
`sec_qm_init()` is only called from `sec_probe()` that is called only when MMU is enabled.